### DISCUSSION TITLE
update + filter type issue

### DISCUSSION THREAD

**Discussion by elsasslegend** [diesel-rs/diesel#3385] [Q&A] [answered]
Hi,

Environment : 
- diesel = { version = "2.0.2", features = ["sqlite"] }
- SQLite on Windows

Creating a schema from the following SQL : 
```
CREATE TABLE "transaction_outbox" (
	"id"	INTEGER PRIMARY KEY NOT NULL,
	"event_data" TEXT NOT NULL,
	"published"	BOOLEAN NOT NULL DEFAULT FALSE
)
```

Generates (the schema) :
```
diesel::table! {
    transaction_outbox (id) {
        id -> Integer,
        event_data -> Text,
        published -> Bool,
    }
}
```

I create the structure : 
```
#[derive(Queryable)]
pub struct TransactionOutbox {
    pub id: i32,
    pub event_data: String,
    pub published: bool,
}
```

Then trying to perform this request : 
```
diesel::update(transaction_outbox.filter((id.eq(item.id), published.eq(false))))
                .set(published.eq(true))
                .execute(self.db.connection());
```

Leads to the following error on .filter method : 

> The trait bound `(diesel::sql_types::Bool, diesel::sql_types::Bool): BoolOrNullableBool` is not satisfied
> the following other types implement trait `BoolOrNullableBool`:
>   Nullable<diesel::sql_types::Bool>
>   diesel::sql_types::Bool
> required because of the requirements on the impl of `FilterDsl<(expression::grouped::Grouped<expression::operators::Eq<schema::transaction_outbox::columns::id, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Integer, i32>>>, expression::grouped::Grouped<expression::operators::Eq<schema::transaction_outbox::columns::published, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Bool, bool>>>)>` for `SelectStatement<FromClause<schema::transaction_outbox::table>>`

Trying to refactor the request to : 
```
diesel::update(transaction_outbox)
                .filter((id.eq(item.id), published.eq(false)))
                .set(published.eq(true))
                .execute(self.db.connection());
```

Produces another error on .filter : 

> The trait bound `(diesel::sql_types::Bool, diesel::sql_types::Bool): BoolOrNullableBool` is not satisfied
> the following other types implement trait `BoolOrNullableBool`:
>   Nullable<diesel::sql_types::Bool>
>   diesel::sql_types::Bool
> required because of the requirements on the impl of `diesel::query_builder::where_clause::WhereAnd<(expression::grouped::Grouped<expression::operators::Eq<schema::transaction_outbox::columns::id, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Integer, i32>>>, expression::grouped::Grouped<expression::operators::Eq<schema::transaction_outbox::columns::published, diesel::internal::derives::as_expression::Bound<diesel::sql_types::Bool, bool>>>)>` for `diesel::query_builder::where_clause::NoWhereClause`

I cannot figure out what is wrong with my schema/model.
Any ideas ?

**Comment by weiznich** [comment, marked as answer]
This is expected behaviour as `.filter()` expects a single boolean sql expression as argument. Tuples of boolean sql expressions are not a boolean sql expression. You might want to use methods like [`BooleanExpressionMethods::and](https://docs.diesel.rs/master/diesel/expression_methods/trait.BoolExpressionMethods.html#method.and) and [`BooleanExpressionMethods::or](https://docs.diesel.rs/master/diesel/expression_methods/trait.BoolExpressionMethods.html#method.and) to combine your conditions.

As another note: I've converted this to a discussion as our issue tracker is only for real issues, not for support questions. In addition such reports need to closely follow the provided template. Please use the discussion forum for future questions.

**Reply by elsasslegend** [reply]
Many thanks this was helpful. I was trying to use the same syntax as for "insert_into", but now I got it.
Sorry for the missing template, it was not proposed at the moment I've created the issue.
